### PR TITLE
Use libvast tree hash for plugin API/ABI checks

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This is a bit of a hack, but it's better than the nothing we've had
before.  Until libvast is stable and versioned, we will have to require
exact version matches between the libvast loading plugins and the
libvast plugins are linking against.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Test the mismatch locally.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,9 @@ Every entry has a category for which we use the following visual abbreviations:
 
 ## Unreleased
 
+- ⚠️ Plugins must now link against the exact same version of libvast that VAST
```
Comment (lava) [inline CHANGELOG.md:15]: Must now be built against [...]

File: libvast/src/plugin.cpp [hunk 2]
```diff
@@ -48,7 +48,10 @@ std::string to_string(plugin_version x) {
 }
 
 bool has_required_version(const plugin_version& version) noexcept {
-  return plugin::version.major == version.major
+  return !std::strcmp(plugin::version.libvast_version, version.libvast_version)
+         && !std::strcmp(plugin::version.libvast_build_tree_hash,
+                         version.libvast_build_tree_hash)
+         && plugin::version.major == version.major
```
Comment (lava) [inline libvast/src/plugin.cpp:54]: Do we still need to compare major/minor/patch/tweak version? I thought we were going to use these as the user-facing plugin version.

File: libvast/vast/plugin.hpp [hunk 3]
```diff
@@ -44,6 +44,14 @@ extern "C" struct plugin_version {
   uint16_t minor;
   uint16_t patch;
   uint16_t tweak;
+
+  /// An additional field containing the version of the libvast that is linked
+  /// against. Should compare equal to VAST_VERSION.
+  const char* libvast_version;
```
Comment (tobim) [inline libvast/vast/plugin.hpp:50]: Why are the preprocessor variables called `VAST_` but the ones here `libvast_*`?

File: libvast/vast/plugin.hpp [hunk 4]
```diff
@@ -44,6 +44,14 @@ extern "C" struct plugin_version {
...
+
+  /// An additional field containing the hash of the build tree of the libvast
+  /// that is linked against. Should compare equal to VAST_BUILD_TREE_HASH.
+  const char* libvast_build_tree_hash;
 };
```
Comment (mavam) [inline libvast/vast/plugin.hpp:55]: These should not go into `plugin_version`, I think. They are conceptually different things and I don't think we need to add them in this particular struct. Can't these just be tracked separate from `plugin_version`?

Comment (dominiklohmann) [reply libvast/vast/plugin.hpp:55]: Then you'll need to add additional exported functions to the plugin API.

File: libvast/vast/plugin.hpp [hunk 5]
```diff
@@ -68,7 +77,8 @@ class plugin {
 public:
   /// The current version of the plugin API. When registering a plugin, set the
   /// corresponding plugin version in the `VAST_REGISTER_PLUGIN` macro.
-  constexpr static auto version = plugin_version{0, 1, 0, 0};
+  constexpr static auto version
+    = plugin_version{0, 1, 0, 0, VAST_VERSION, VAST_BUILD_TREE_HASH};
```
Comment (tobim) [inline libvast/vast/plugin.hpp:81]: Do we still need the `major` `minor` `patch` and `tweak` members in this struct? I guess it makes sense when we have some stability guarantees?

**Review by tobim — COMMENTED** [review]
More questions than change requests.

File: CMakeLists.txt [hunk 6]
```diff
@@ -36,6 +36,28 @@ if (NOT IS_ABSOLUTE "${CMAKE_INSTALL_PREFIX}")
       "CMAKE_INSTALL_PREFIX must be an absolute path: ${CMAKE_INSTALL_PREFIX}")
 endif ()
 
+# -- build tree hash -----------------------------------------------------------
+
+file(GLOB_RECURSE hash_files CONFIGURE_DEPENDS "libvast/*" "cmake/*"
+     "CMakeLists.txt")
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply CMakeLists.txt:42]: That's already happening directly below:

```
list(FILTER hash_files EXCLUDE REGEX
     "^${CMAKE_CURRENT_SOURCE_DIR}/libvast/aux/")
```

File: libvast/vast/plugin.hpp (hunk 3 above)
Comment (dominiklohmann) [reply libvast/vast/plugin.hpp:50]: In terms of the plugin, it is libvast that is relevant for us. Within libvast, all macros are prefix with `VAST_`.

File: libvast/vast/plugin.hpp (hunk 5 above)
Comment (dominiklohmann) [reply libvast/vast/plugin.hpp:81]: For now I've opted to keep them, although they're pretty much useless until we make such guarantees. Shall I remove them for now?

File: libvast/vast/plugin.hpp (hunk 4 above)
Comment (mavam) [reply libvast/vast/plugin.hpp:55]: Seems cleaner to me that way, now that you mention it.

File: libvast/src/plugin.cpp (hunk 2 above)
Comment (dominiklohmann) [reply libvast/src/plugin.cpp:54]: We can always change this later on.

File: libvast/src/plugin.cpp [hunk 7]
```diff
@@ -59,17 +59,30 @@ bool has_required_version(const plugin_version& version) noexcept {
...
-      auto plugin_create = reinterpret_cast<::vast::plugin* (*) ()>(
-        dlsym(library_, "vast_plugin_create"));
-      auto plugin_destroy = reinterpret_cast<void (*)(::vast::plugin*)>(
-        dlsym(library_, "vast_plugin_destroy"));
-      if (plugin_create && plugin_destroy) {
-        instance_ = plugin_create();
-        deleter_ = plugin_destroy;
-      }
+    if (!plugin_version || !has_required_version(plugin_version()))
+      return;
```
Comment (lava) [inline libvast/src/plugin.cpp:77]: We should print some kind of appropriate error message if loading the plugin fails. (We can also do this in a follow-up PR if you keep working on plugins anyways.)

Comment (dominiklohmann) [reply libvast/src/plugin.cpp:77]: Yeah I had the same thought already, probably moving from `plugin_ptr::plugin_ptr(...)` to `caf::expected<plugin_ptr> plugin_ptr::make(...)`.

**Review by lava — APPROVED** [review]
Not tested personally, but looks good to me and reflects our discussions (and since its a new feature there's no chance of regressions :)

**Pull request merged**